crates/sui-core/src/execution_engine.rs

```diff
@@ -380,63 +412,24 @@ fn pay<S>(
         ));
     }
 
-    // walk through the coins from left to right, debiting as needed to cover each amount to send
-    let mut cur_coin_idx = 0;
-    for (recipient, amount) in recipients.iter().zip(amounts) {
-        let mut remaining_amount = amount;
-        loop {
-            // while remaining_amount != 0
-            // guaranteed to be in-bounds because of the total > total_coins check above
-            let coin = &mut coins[cur_coin_idx];
-            let coin_value = coin.value();
-            if coin_value == 0 {
-                // can't extract any more value from this coin, go to the next one
-                cur_coin_idx += 1;
-                continue;
-            }
-            if coin_value >= remaining_amount {
-                // can get everything we need from this coin
-                coin.balance.withdraw(remaining_amount).unwrap();
-                // create a new coin for the recipient with the original amount
-                let new_coin = Object::new_move(
-                    MoveObject::new_coin(
-                        coin_type.clone(),
-                        OBJECT_START_VERSION,
-                        bcs::to_bytes(&Coin::new(UID::new(tx_ctx.fresh_id()), amount))
-                            .expect("Serializing coin value cannot fail"),
-                    ),
-                    Owner::AddressOwner(*recipient),
-                    tx_ctx.digest(),
-                );
-                temporary_store.write_object(new_coin, WriteKind::Create);
-                break; // done paying this recipieint, on to the next one
-            } else {
-                // need to take all of this coin and some from a subsequent coin
-                coin.balance.withdraw(coin_value).unwrap();
-                remaining_amount -= coin_value;
-            }
-        }
-    }
-
-    #[cfg(debug_assertions)]
-    {
-        // double check that we didn't create or destroy money
-        let new_total_coins = coins.iter().fold(0, |acc, c| acc + c.value());
-        assert_eq!(total_coins - new_total_coins, total_amount)
-    }
-
-    // update the input coins to reflect the decrease in value.
-    // if the input coin has value 0, delete it
+    let mut remaining_coin = total_coins;
```

Comment:
the main change here is, instead of creating new coins on L402 and deleting input coins on L433, here I tried to reuse the input coins and transfer them to recipients with specified amounts.